File: employee_self_service/employee_self_service/page/target_allocation_tool/target_allocation_tool.py
import frappe
from frappe.utils import getdate, flt
from frappe import _


@frappe.whitelist()
def get_filtered_targets(employee, from_date, to_date):
    """Get Employee Target Entries filtered by employee and date range."""
    if not employee or not from_date or not to_date:
        return []

    filters = {
        "employee": employee,
        "start_date": ["<=", to_date],
        "end_date": [">=", from_date],
    }

    targets = frappe.get_all(
        "Employee Target Entry",
        filters=filters,
        fields=[
            "name",
            "target_template",
            "start_date",
            "end_date",
            "total_target",
            "total_achieved",
            "status",
            "progress",
        ],
        order_by="creation desc",
    )

    return targets


@frappe.whitelist()
def get_sales_transactions(employee, from_date, to_date, target_entry=None):
    """Get sales orders and invoices for the employee within date range based on target entry."""
    if not employee or not from_date or not to_date or not target_entry:
        frappe.throw(
            _(
                "All parameters are required: Employee, From Date, To Date, and Target Entry"
            )
        )

    # Get sales person from employee
    sales_person = frappe.db.get_value("Sales Person", {"employee": employee}, "name")

    if not sales_person:
        frappe.throw(_("No Sales Person found for Employee {0}").format(employee))

    # Get target entry details to determine which documents to fetch
    target_doc = frappe.get_doc("Employee Target Entry", target_entry)
    target_template = frappe.get_doc(
        "Employee Target Template", target_doc.target_template
    )

    transactions = []

    # Determine which document types to fetch based on target module
    if target_template.target_module == "Sales Order":
        # Fetch only Sales Orders
        transactions = _get_sales_orders(sales_person, from_date, to_date)
    elif target_template.target_module == "Sales Invoice":
        # Fetch only Sales Invoices
        transactions = _get_sales_invoices(sales_person, from_date, to_date)
    else:
        # Default to both if module is not specified or other value
        sales_orders = _get_sales_orders(sales_person, from_date, to_date)
        sales_invoices = _get_sales_invoices(sales_person, from_date, to_date)
        transactions = sales_orders + sales_invoices

    # Check allocation status for each transaction
    for txn in transactions:
        # Check if already allocated
        existing_log = frappe.db.get_value(
            "SP Target Log",
            {
                "reference_docname": txn["name"],
                "reference_doctype": txn["doctype"],
                "docstatus": 1,
            },
            ["name", "employee_target_entry"],
        )

        if existing_log:
            txn["is_allocated"] = True
            txn["allocated_target"] = (
                existing_log[1] if isinstance(existing_log, tuple) else existing_log
            )
            txn["target_log"] = (
                existing_log[0] if isinstance(existing_log, tuple) else existing_log
            )
        else:
            txn["is_allocated"] = False

    return transactions


def _get_sales_orders(sales_person, from_date, to_date):
    """Get sales orders for the sales person."""
    sales_orders = frappe.db.sql(
        """
        SELECT DISTINCT 
            so.name,
            so.customer,
            so.customer_name,
            so.transaction_date,
            so.grand_total,
            'Sales Order' as doctype,
            so.status
        FROM `tabSales Order` so
        INNER JOIN `tabSales Team` st ON st.parent = so.name
        WHERE st.sales_person = %s
        AND so.docstatus = 1
        AND so.transaction_date BETWEEN %s AND %s
        ORDER BY so.transaction_date DESC
    """,
        (sales_person, from_date, to_date),
        as_dict=True,
    )

    return sales_orders


def _get_sales_invoices(sales_person, from_date, to_date):
    """Get sales invoices for the sales person."""
    sales_invoices = frappe.db.sql(
        """
        SELECT DISTINCT 
            si.name,
            si.customer,
            si.customer_name,
            si.posting_date as transaction_date,
            si.grand_total,
            'Sales Invoice' as doctype,
            si.status
        FROM `tabSales Invoice` si
        INNER JOIN `tabSales Team` st ON st.parent = si.name
        WHERE st.sales_person = %s
        AND si.docstatus = 1
        AND si.posting_date BETWEEN %s AND %s
        ORDER BY si.posting_date DESC
    """,
        (sales_person, from_date, to_date),
        as_dict=True,
    )

    return sales_invoices


@frappe.whitelist()
def allocate_transaction_to_target(employee, target_entry, doc_type, doc_name):
    """Create SP Target Log entry for the transaction."""

    # Validate inputs
    if not all([employee, target_entry, doc_type, doc_name]):
        frappe.throw(_("All parameters are required"))

    # Check if already allocated
    existing_log = frappe.db.exists(
        "SP Target Log",
        {"reference_docname": doc_name, "reference_doctype": doc_type, "docstatus": 1},
    )

    if existing_log:
        frappe.throw(
            _("{0} {1} is already allocated to a target").format(doc_type, doc_name)
        )

    # Get document details
    doc = frappe.get_doc(doc_type, doc_name)
    target_doc = frappe.get_doc("Employee Target Entry", target_entry)

    # Create SP Target Log
    target_log = frappe.new_doc("SP Target Log")
    target_log.employee = employee
    target_log.employee_target_entry = target_entry
    target_log.customer = doc.customer
    target_log.customer_name = doc.customer_name
    target_log.reference_doctype = doc_type
    target_log.reference_docname = doc_name
    target_log.amount = doc.grand_total
    target_log.transaction_date = (
        doc.transaction_date if doc_type == "Sales Order" else doc.posting_date
    )
    target_log.date = frappe.utils.today()
    target_log.metric = target_doc.metric

    # Get customer group
    customer_group = frappe.db.get_value("Customer", doc.customer, "customer_group")
    if customer_group:
        target_log.customer_group = customer_group

    # Calculate quantity if needed for quantity-based targets
    if target_doc.metric == "Quantity":
        total_qty = (
            frappe.db.sql(
                """
            SELECT SUM(qty) FROM `tab{} Item` WHERE parent = %s
        """.format(
                    doc_type
                ),
                (doc_name,),
            )[0][0]
            or 0
        )
        target_log.qty = total_qty

    # Get item group from first item if needed
    if target_doc.selector == "Item Group":
        item_group = frappe.db.sql(
            """
            SELECT i.item_group FROM `tab{} Item` di
            INNER JOIN `tabItem` i ON i.name = di.item_code
            WHERE di.parent = %s
            LIMIT 1
        """.format(
                doc_type
            ),
            (doc_name,),
        )
        if item_group:
            target_log.item_group = item_group[0][0]

    try:
        target_log.insert()
        target_log.submit()
        return {
            "success": True,
            "message": _("Successfully allocated {0} {1} to target").format(
                doc_type, doc_name
            ),
            "target_log": target_log.name,
        }
    except Exception as e:
        frappe.log_error(f"Error creating SP Target Log: {str(e)}")
        frappe.throw(_("Error allocating transaction: {0}").format(str(e)))


@frappe.whitelist()
def unallocate_transaction(doc_type, doc_name):
    """Cancel SP Target Log entry for the transaction to unallocate it."""

    # Validate inputs
    if not all([doc_type, doc_name]):
        frappe.throw(_("Document Type and Document Name are required"))

    # Find the existing target log
    existing_log = frappe.db.get_value(
        "SP Target Log",
        {"reference_docname": doc_name, "reference_doctype": doc_type, "docstatus": 1},
        ["name", "employee_target_entry"],
    )

    if not existing_log:
        frappe.throw(
            _("{0} {1} is not allocated to any target").format(doc_type, doc_name)
        )

    try:
        # Get the target log document
        target_log_name = (
            existing_log[0] if isinstance(existing_log, tuple) else existing_log
        )
        target_log = frappe.get_doc("SP Target Log", target_log_name)

        # Cancel the target log
        target_log.cancel()

        return {
            "success": True,
            "message": _("Successfully unallocated {0} {1} from target").format(
                doc_type, doc_name
            ),
            "cancelled_log": target_log_name,
        }
    except Exception as e:
        frappe.log_error(f"Error cancelling SP Target Log: {str(e)}")
        frappe.throw(_("Error unallocating transaction: {0}").format(str(e)))


@frappe.whitelist()
def get_employee_list():
    """Get list of employees who have sales person linked."""
    employees = frappe.db.sql(
        """
        SELECT DISTINCT e.name, e.employee_name
        FROM `tabEmployee` e
        INNER JOIN `tabSales Person` sp ON sp.employee = e.name
        WHERE e.status = 'Active'
        ORDER BY e.employee_name
    """,
        as_dict=True,
    )

    return employees

File: employee_self_service/employee_self_service/page/target_allocation_tool/target_allocation_tool.js
frappe.pages['target-allocation-tool'].on_page_load = function (wrapper) {
    var page = frappe.ui.make_app_page({
        parent: wrapper,
        title: __('Target Allocation Tool'),
        single_column: true
    });

    // Add filter fields to the page
    page.from_date = page.add_field({
        fieldname: 'from_date',
        label: __('From Date'),
        fieldtype: 'Date',
        default: frappe.datetime.month_start(),
        reqd: 1,
        change: function () {
            page.target_entry.set_value('');
            page.clear_results();
        }
    });

    page.to_date = page.add_field({
        fieldname: 'to_date',
        label: __('To Date'),
        fieldtype: 'Date',
        default: frappe.datetime.get_today(),
        reqd: 1,
        change: function () {
            page.target_entry.set_value('');
            page.clear_results();
        }
    });

    page.employee = page.add_field({
        fieldname: 'employee',
        label: __('Employee'),
        fieldtype: 'Link',
        options: 'Employee',
        reqd: 1,
        get_query: function () {
            return {
                filters: {
                    'status': 'Active'
                }
            };
        },
        change: function () {
            page.target_entry.set_value('');
            page.update_target_filter();
            page.clear_results();
        }
    });

    page.target_entry = page.add_field({
        fieldname: 'target_entry',
        label: __('Target Entry'),
        fieldtype: 'Link',
        options: 'Employee Target Entry',
        reqd: 1,
        get_query: function () {
            if (page.employee.get_value() && page.from_date.get_value() && page.to_date.get_value()) {
                return {
                    filters: {
                        'employee': page.employee.get_value(),
                        'start_date': ['<=', page.to_date.get_value()],
                        'end_date': ['>=', page.from_date.get_value()]
                    }
                };
            }
            return {};
        },
        change: function () {
            page.clear_results();
        }
    });

    // Add Get Details button
    page.add_inner_button(__('Get Transaction Details'), function () {
        page.get_transaction_details();
    }).addClass('btn-primary');

    // Helper functions
    page.update_target_filter = function () {
        if (page.target_entry) {
            page.target_entry.refresh();
        }
    };

    page.clear_results = function () {
        page.main.find('.transaction-results').remove();
    };

    page.get_transaction_details = function () {
        if (!page.employee.get_value() || !page.from_date.get_value() || !page.to_date.get_value() || !page.target_entry.get_value()) {
            frappe.msgprint(__('Please fill all mandatory fields: Employee, From Date, To Date, and Target Entry'));
            return;
        }

        frappe.call({
            method: 'employee_self_service.employee_self_service.page.target_allocation_tool.target_allocation_tool.get_sales_transactions',
            args: {
                employee: page.employee.get_value(),
                from_date: page.from_date.get_value(),
                to_date: page.to_date.get_value(),
                target_entry: page.target_entry.get_value()
            },
            callback: function (r) {
                if (r.message) {
                    page.render_transaction_table(r.message);
                }
            }
        });
    };

    page.render_transaction_table = function (data) {
        page.clear_results();

        if (!data || data.length === 0) {
            $(`<div class="transaction-results">
                <div class="alert alert-info">
                    <strong>No transactions found</strong> for the selected criteria.
                </div>
            </div>`).appendTo(page.main);
            return;
        }

        let html = `
            <div class="transaction-results" style="margin-top: 20px;">
                <h5>Transaction Details</h5>
                <div class="table-responsive">
                    <table class="table table-bordered table-striped">
                        <thead>
                            <tr>
                                <th>Document Type</th>
                                <th>Document No</th>
                                <th>Date</th>
                                <th>Customer</th>
                                <th>Total Amount</th>
                                <th>Status</th>
                                <th>Action</th>
                            </tr>
                        </thead>
                        <tbody>
        `;

        data.forEach(function (row) {
            let statusBadge = '';
            let actionButton = '';

            if (row.is_allocated) {
                statusBadge = '<span class="badge badge-success">Allocated</span>';
                actionButton = `<button class="btn btn-sm btn-warning unallocate-btn"
                                       data-doctype="${row.doctype}"
                                       data-docname="${row.name}"
                                       data-target-log="${row.allocated_target || ''}">
                                   Unallocate
                               </button>`;
            } else {
                statusBadge = '<span class="badge badge-warning">Not Allocated</span>';
                actionButton = `<button class="btn btn-sm btn-primary allocate-btn"
                                       data-doctype="${row.doctype}"
                                       data-docname="${row.name}">
                                   Allocate
                               </button>`;
            }

            html += `
                <tr>
                    <td>${row.doctype}</td>
                    <td><a href="/app/${row.doctype.toLowerCase().replace(' ', '-')}/${row.name}" target="_blank">${row.name}</a></td>
                    <td>${frappe.datetime.str_to_user(row.transaction_date)}</td>
                    <td>${row.customer_name}</td>
                    <td>${format_currency(row.grand_total)}</td>
                    <td>${statusBadge}</td>
                    <td>${actionButton}</td>
                </tr>
            `;
        });

        html += `
                        </tbody>
                    </table>
                </div>
            </div>
        `;

        $(html).appendTo(page.main);

        // Bind allocation button events
        page.main.find('.allocate-btn').on('click', function () {
            let doctype = $(this).data('doctype');
            let docname = $(this).data('docname');
            page.allocate_transaction(doctype, docname);
        });

        // Bind unallocation button events
        page.main.find('.unallocate-btn').on('click', function () {
            let doctype = $(this).data('doctype');
            let docname = $(this).data('docname');
            let targetLog = $(this).data('target-log');
            page.unallocate_transaction(doctype, docname, targetLog);
        });
    };

    page.allocate_transaction = function (doctype, docname) {
        if (!page.target_entry.get_value()) {
            frappe.msgprint(__('Please select a Target Entry first'));
            return;
        }

        frappe.confirm(
            __('Are you sure you want to allocate {0} <strong>{1}</strong> to the selected target?', [doctype, docname]),
            function () {
                frappe.call({
                    method: 'employee_self_service.employee_self_service.page.target_allocation_tool.target_allocation_tool.allocate_transaction_to_target',
                    args: {
                        employee: page.employee.get_value(),
                        target_entry: page.target_entry.get_value(),
                        doc_type: doctype,
                        doc_name: docname
                    },
                    callback: function (r) {
                        if (r.message && r.message.success) {
                            frappe.show_alert({
                                message: r.message.message,
                                indicator: 'green'
                            });
                            // Refresh the table
                            page.get_transaction_details();
                        }
                    }
                });
            }
        );
    };

    page.unallocate_transaction = function (doctype, docname, targetLog) {
        frappe.confirm(
            __('Are you sure you want to unallocate {0} <strong>{1}</strong>? This will cancel the target log.', [doctype, docname]),
            function () {
                frappe.call({
                    method: 'employee_self_service.employee_self_service.page.target_allocation_tool.target_allocation_tool.unallocate_transaction',
                    args: {
                        doc_type: doctype,
                        doc_name: docname
                    },
                    callback: function (r) {
                        if (r.message && r.message.success) {
                            frappe.show_alert({
                                message: r.message.message,
                                indicator: 'orange'
                            });
                            // Refresh the table
                            page.get_transaction_details();
                        }
                    }
                });
            }
        );
    };
};